nestris_ocr/scan_strat/scan_helpers.py

```diff
@@ -80,26 +103,90 @@ def scan_lines(full_image, digit_mask="OOO"):
     return scan_text(full_image, PATTERNS["lines"], digit_mask, WINDOW_AREAS["lines"])
 
 
-def scan_field(full_image, color1=None, color2=None):
-    sub_image = get_sub_image(full_image, WINDOW_AREAS["field"])
+def scan_colors(full_image):
+    color1 = get_sub_image(full_image, WINDOW_AREAS["color1"])
+    color1 = color1.resize((1, 1), PIL.Image.ANTIALIAS)
+    color1 = color1.getpixel((0, 0))
+    color1 = np.array(color1, dtype=np.uint8)
+
+    color2 = get_sub_image(full_image, WINDOW_AREAS["color2"])
+    color2 = color2.resize((1, 1), PIL.Image.ANTIALIAS)
+    color2 = color2.getpixel((0, 0))
+    color2 = np.array(color2, dtype=np.uint8)
+
+    return {"color1": color1, "color2": color2}
```

Comment:
/nit I see you have a lot of painful usages of:
```python
self.color1 = result["color1"]
self.color2 = result["color2"]
```

That's the result of dict not being treated as class attribute (unlike Javascript). Maybe returning a tuple here would be easier? `return color1, color2`

So the code everywhere can be:
```python
self.color1, self.color2 = result
```

Reply:
Yeah originally it was tuple, then i changed it to dict, and now i think maybe tuple is the way to go.
I changed it to dict because it was returning 3 things at some point, and i think once you get past 2 items, returning a big tuple is less clean than a result dict.

Or i guess maybe we should be passing around a ColorTable() object that has all four colors, and the function can update the ColorTable?

Reply:
It's fine then. We can stay with `dict` for now and refactor later. Not introducing divergence would be more important and should be easier to modify in the future 😃

Reply:
Yeah lets refactor to some form of object that stores all four colors in the future.

Reply:
Aye 😅 .  did initially refactor to tuples, but reverted to dictionary because I wanted the `scan_color()` and `scan_bnw()` to have the same return shape, but black and white are more complicated structures (storing the rgb tuple and the luma value) which "felt" a little weird in a tuple return values.

In the end that rationale is not super valid I guess 😅 

>  Not introducing divergence would be more important 

I don't think we have to worry too much about this, considering the PR completely changes how the color are read, and which component is responsible for scanning them. 

Scanning the colors used to be an embedded task in the field scanner, which, alongside the field data, returned color1 and color2 (which the strategy could then cache). Now the strategy is in charge of scanning colors explicitely (and deciding when to cache), and the colors must always be supplied to the field scanner.

I like the idea of a `ColorTable` object that can be updated by different scanning tasks. Would make things a lot easier than having `black, white, color1, color2` in function signatures all over the place.

Although it does mean I must find a better way to precompute the black and white luma and store it. I suppose it could be as simple s having the `ColorTable` object having the field:
```
{
  black,
  white,
  color1,
  color2,
  black_luma,
  white_luma
}
```

I'll do that refactoring and add commit to the PR. Update coming soon! 😄

Reply:
Sounds like what you actually want might be a `namedtuple`: https://docs.python.org/3/library/collections.html#collections.namedtuple

It combines the benefits of tuple and attribute based access, very flexible and very simple.

For now you can just throw the type definition in https://github.com/alex-ong/NESTrisOCR/blob/master/nestris_ocr/types.py

Reply:
Will do! Thanks Blake :)

Reply:
I added a commit to clean this up!

Having the object `colors` makes much of the code cleaner. Also it's a logical place to place the transformations to `np.array(dtype=np.uint8)`, so I could remove that from the scan_helpers, and it meant `scan_color()` and `scan_bnw()` can have the same shape, and a tuple return value makes sense now.

Passing `colors` around also allows the scanners themselves to pick whatever they need from the structure.

I did not use a named tuple inthe end because namedtuple must be initialized for all its field at construction, but in the code:
* black/white luma are internal derived values
* black/white and color1/color2 are typically added at different times

The result still looks much cleaner now. 

One last thing bothers me: it feels weird my class `Colors` is implemented in `types.py`. Should I extract this to its own file instead?

Reply:
yes, and you sure did.

Reply:
Haha yes! Now it's a proper class then definitely is worthy its own file 👍